launch/launch/utilities/type_utils.py

```diff
@@ -0,0 +1,565 @@
+
+
+def get_typed_value(
+    value: Union[Text, List[Text]],
+    data_type: Optional[AllowedTypesType],
+    can_be_str: bool = False,
+) -> StrSomeValueType:
+    """
+    Try to convert `value` to the type specified in `data_type`.
+
+    :param value: string or list of strings to be coerced.
+    :param data_type: value will be coerced to data_type.
+    :param can_be_str: if `True`, strings will be kept in case coercion fails.
+      In the case of lists, it will also accepts strings as items.
+      launch.frontend makes use of this for string embedded substitutions.
+    :raise: `ValueError` if the coercion failed.
+    :raise: `TypeError` if `value` is a `list` and `data_type` is not a `typing.List[x]` object.
+    :raise: `TypeError` if `value` is neither a `str` of a list of `str`.
+    :return: `value` coerced to `data_type`.
+    """
+    if isinstance(value, list):
+        if data_type is not None:
+            data_type, is_list = extract_type(data_type)
+            if not is_list:
+                raise TypeError(
+                    f"Cannot convert input '{value}' of type '{type(value)}' to"
+                    f" '{data_type}'"
+                )
+        output: AllowedValueType = coerce_list(value, data_type, can_be_str)
+    else:
+        output = coerce_to_type(value, data_type, can_be_str)
+    return output
+
+
+def is_substitution(x):
+    """
+    Return `True` if `x` is some substitution.
+
+    This can be:
+    - A :py:class:`launch.Substitution` instance.
+    - A list of mixed `launch.Substitution` and `str` instances,
+        with at least one instance of the former.
+    """
+    return (
+        isinstance(x, Substitution) or
+        (
+            isinstance(x, collections.abc.Iterable) and
+            len(x) > 0 and
+            all(isinstance(item, (Substitution, str)) for item in x) and
+            any(isinstance(item, Substitution) for item in x)
+        )
+    )
+
+
+def normalize_typed_substitution(
+    value: SomeValueType, data_type: Optional[AllowedTypesType]
+) -> NormalizedValueType:
+    """
+    Normalize a mix of substitution and values.
+
+    This function becomes handy when you need a substitution which result will be coerced to a
+    specific type.
+
+    Example:
+    -------
+    ```python3
+    class MyAction(Action):
+        def __init__(self, my_int: Union[int, SomeSubstitutionsType]):
+            self.__my_int_normalized = normalize_typed_substitution(some_int, int)
+            ...
+
+        def execute(self, context):
+            my_int = perform_typed_substitution(context, self.__my_int_normalized, int)
+            ...
+    ```
+
+    :param value: value to be normalized.
+    :param data_type: `value` can be either an instance of `data_type` or a substitution.
+        In the case of lists, `value` can be either a substitution or a list.
+        In the later case, its item should match the type specified by `data_type` or be a
```

Comment:
@ivanpauno typo: 
```suggestion
        In the later case, its items should match the type specified by `data_type` or be a
```

Reply:
Fixed in https://github.com/ros2/launch/pull/438/commits/cd3fafeca6102d69dc480c3a887ea74a2e659e8a.